### PR TITLE
Add function_data classes from InfrastructureSystems

### PR DESCRIPTION
**Pull request by jerrypotts**
Added new file (function_data.py) to include the function_data structs from InfrastructureSystems as python classes, which are necessary for parsing data for cost function inputs into Sienna. I included classes for linear and quadratic functions, as well as piecewise linear and piecewise step functions.

Note, we kept PiecewiseLinearData.points as a list of NamedTuples (rather than change it to two separate lists of x, y coordinates) to ensure it is consistent with the structs built on top of function_data in InfrastructureSystems and elsewhere.

### CODE REVIEW

File: src/infrasys/function_data.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    represent the derivative of a [PiecewiseLinearData](@ref), where the y-coordinates of this
+    step function represent the slopes of that piecewise linear function, so there is also an
+    optional field `c` that can be used to store the initial y-value of that piecewise linear
+    function. Principally used for the representation of cost functions where the points store
+    quantities (x, dy/dx), such as (MW, \$/MWh).
+    """
+
+    name: Annotated[str, Field(frozen=True)] = ""
+    x_coords: Annotated[
+        list[float], Field(description="the x-coordinates of the endpoints of the segments")
```
Comment (GabrielKS) [inline src/infrasys/function_data.py:97]: In the Julia version, these `x_coords` get validated the same way as in `PiecewiseLinearData`. Also we validate that `len(y_coords) == len(x_coords) - 1`

File: src/infrasys/function_data.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class PiecewiseStepData(Component):
+    r"""
+    Class to represent a step function as a series of endpoint x-coordinates and segment
+    y-coordinates: two x-coordinates and one y-coordinate defines a single segment, three
+    x-coordinates and two y-coordinates define two segments, etc. This can be useful to
+    represent the derivative of a [PiecewiseLinearData](@ref), where the y-coordinates of this
+    step function represent the slopes of that piecewise linear function, so there is also an
+    optional field `c` that can be used to store the initial y-value of that piecewise linear
```
Comment (GabrielKS) [inline src/infrasys/function_data.py:90]: the stuff about the optional field `c` is out of date, that's my fault for not removing it from the Julia version. Will be done over there in the next few days and then you can re-copypaste.

**Review by GabrielKS — CHANGES_REQUESTED** [review]

File: src/infrasys/function_data.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    ]
+    proportional_term: Annotated[
+        float, Field(description="the proportional term in the represented function")
+    ]
+    constant_term: Annotated[
+        float, Field(description="the constant term in the represented function")
+    ]
+
+
+def validate_piecewise_x(points: list):
```
Comment (pesap) [inline src/infrasys/function_data.py:49]: Add type hints.
```suggestion
def validate_piecewise_x(points: list[NamedTuple]) -> list[NamedTuple]:
```

Comment (pesap) [inline src/infrasys/function_data.py:49]: Could you add docstring to the function? See this example:  https://numpydoc.readthedocs.io/en/latest/example.html

File: src/infrasys/function_data.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    if isinstance(points[0], XY_COORDS):
+        x_coords = [p.x for p in points]
+    else:
+        x_coords = points
```
Comment (pesap) [inline src/infrasys/function_data.py:53]: By looking at. this it looks like  points can be either a named tuple of a list of floats? If so, could you update the type hint of the function.

File: src/infrasys/function_data.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        x_coords == sorted(x_coords)
+        or (np.isnan(x_coords[0]) and x_coords[1:] == sorted(x_coords[1:]))
+    ):
+        raise ValueError(f"Piecewise x-coordinates must be ascending, got {x_coords}")
+
+    return points
+
+
+class PiecewiseLinearData(Component):
+    r"""
```
Comment (pesap) [inline src/infrasys/function_data.py:67]: r not needed

File: tests/test_function_data.py
Comment (pesap) [inline tests/test_function_data.py:1]: Can you add a function that checks that `XY_COORDS`?
An example would be to initialize it, check that is an instance of XY_COORd and check the x and y value are setup correctly.

**Review by pesap — CHANGES_REQUESTED** [review]
LGTM, can you address the changes and update the docstring to match the following style: https://numpydoc.readthedocs.io/en/latest/example.html.

File: src/infrasys/function_data.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
+from infrasys import Component
+from typing_extensions import Annotated
+from pydantic import Field, model_validator
+from pydantic.functional_validators import AfterValidator
+from typing import NamedTuple
+import numpy as np
+
+
+class XY_COORDS(NamedTuple):
```
Comment (daniel-thom) [inline src/infrasys/function_data.py:9]: Can this be renamed using PascalCase, such as `XYCoords`? Maybe that is a question for @GabrielKS on the Julia side.

Comment (pesap) [reply src/infrasys/function_data.py:11]: @GabrielKS, any reactions to this? Can we renamed it on the Python version?

Comment (GabrielKS) [reply src/infrasys/function_data.py:11]: The rationale for `XY_COORDS` on the Julia side is that it's a `const` type alias:
```julia
const XY_COORDS = @NamedTuple{x::Float64, y::Float64}
```
and that's how we did those in the past in IS:
```julia
const CONSTANT = Float64  # still exists
const POLYNOMIAL = Tuple{Float64, Float64}  # superseded by FunctionData
const PWL = Vector{Tuple{Float64, Float64}}  # superseded by FunctionData
```

. I'm not opposed to `XYCoords` in the Python version, and there is even an argument for renaming it in the Julia version, given that in PSY we have
```julia
const MinMax = NamedTuple{(:min, :max), Tuple{Float64, Float64}}  # still exists
const UpDown = NamedTuple{(:up, :down), Tuple{Float64, Float64}}  # still exists
```
etc.

**Comment by pesap** [comment]
@jerrypotts , we are working on a patch for the runner so that this can be tested. Let's leave the update of the fields to a following PR.

**Pull request merged**